src/libstd/sys/unix/thread.rs

```diff
@@ -301,7 +270,47 @@ pub unsafe fn detach(native: rust_thread) {
     assert_eq!(pthread_detach(native), 0);
 }
 
-pub unsafe fn yield_now() { assert_eq!(sched_yield(), 0); }
+pub unsafe fn yield_now() {
+    assert_eq!(sched_yield(), 0);
+}
+
+pub fn sleep(dur: Duration) {
+    unsafe {
+        if dur < Duration::zero() {
+            return yield_now()
+        }
+        let seconds = dur.num_seconds();
+        let ns = dur - Duration::seconds(seconds);
+        let mut ts = libc::timespec {
+            tv_sec: seconds as libc::time_t,
+            tv_nsec: ns.num_nanoseconds().unwrap() as libc::c_long,
+        };
+        // If we're awoken with a signal then the return value will be -1 and
+        // nanosleep will fill in `ts` with the remaining time.
+        while libc::nanosleep(&ts, &mut ts) == -1 {
+            assert_eq!(os::errno(), libc::EINTR);
+        }
+    }
+}
+
+pub fn sleep(dur: Duration) {
+    unsafe {
+        if dur < Duration::zero() {
+            return yield_now()
+        }
+        let seconds = dur.num_seconds();
+        let ns = dur - Duration::seconds(seconds);
+        let mut ts = libc::timespec {
+            tv_sec: seconds as libc::time_t,
+            tv_nsec: ns.num_nanoseconds().unwrap() as libc::c_long,
+        };
+        // If we're awoken with a signal then the return value will be -1 and
+        // nanosleep will fill in `ts` with the remaining time.
+        while libc::nanosleep(&ts, &mut ts) == -1 {
+            assert_eq!(os::errno(), libc::EINTR);
+        }
```

Comment:
This should use `clock_nanosleep` to prevent the problems you're trying to work around here:

```
       The fact that nanosleep() sleeps for a relative interval can be	problematic  if	 the
       call  is	 repeatedly  restarted	after  being  interrupted by signals, since the time
       between the interruptions and restarts of the call will lead to	drift  in  the	time
       when   the   sleep   finally  completes.	  This	problem	 can  be  avoided  by  using
       clock_nanosleep(2) with an absolute time value.
```